### ISSUE TITLE
false positive on reverse() of custom class implementing Sequence protocol

### ISSUE THREAD

**Issue by mrwonko** [python/mypy#11233]
**Bug Report**

I have a custom class that implements `__len__` and `__getitem__` as described in [the documentation of `reversed`](https://docs.python.org/3.8/library/functions.html#reversed), thereby fulfilling the `Sequence` protocol. mypy complains that it is not a valid argument to `reversed`.

**To Reproduce**

```Python
class CustomReversible:
    def __init__(self) -> None:
        self.__data = ["foo", "bar"]

    def __len__(self) -> int:
        return len(self.__data)

    def __getitem__(self, item: int) -> str:
        return self.__data[item]


print(list(reversed(CustomReversible())))
```

**Expected Behavior**

```txt
Success: no issues found in 1 source file
```

**Actual Behavior**

```txt
mypy_repro.py:12: error: No overload variant of "reversed" matches argument type "CustomReversible"  [call-overload]
    print(list(reversed(CustomReversible())))
               ^
mypy_repro.py:12: note: Possible overload variants:
mypy_repro.py:12: note:     def [_T] reversed(self, Sequence[_T]) -> reversed[_T]
mypy_repro.py:12: note:     def [_T] reversed(self, Reversible[_T]) -> reversed[_T]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.8.2
- Operating system and version: macOS Big Sur 11.6

**Comment by mrwonko** [comment]
My real code also adds TypeVars into the mix, something close to this:

```python
from typing import List, Sequence, TypeVar, Union, overload

T = TypeVar("T")


class CustomReversible(Sequence[T]):
    def __init__(self, data: List[T]) -> None:
        self.__data = data

    def __len__(self) -> int:
        return len(self.__data)

    @overload
    def __getitem__(self, index: int) -> T:
        ...

    @overload
    def __getitem__(self, index: slice) -> Sequence[T]:
        ...

    def __getitem__(self, index: Union[int, slice]) -> Union[T, Sequence[T]]:
        if isinstance(index, slice):
            raise NotImplementedError("only int index supported")
        else:
            return self.__data[index]


print(list(reversed(CustomReversible(["foo", "bar"]))))
```

Is there some way to shorten that? If I don't explicitly define the overloads, mypy complains
```txt
mypy_repro.py:13: error: Signature of "__getitem__" incompatible with supertype "Sequence"  [override]
        def __getitem__(self, index: Union[int, slice]) -> Union[T, Sequence[T]]:
        ^
Found 1 error in 1 file (checked 1 source file)
```

**Comment by mrwonko** [comment]
In hindsight, using `Sequence[str]` as an example may have been a bad idea, because `str` is a `Sequence[str]` itself…

**Comment by mrwonko** [comment]
`Sequence` is curiously absent from the [list of protocols in the mypy documentation](https://mypy.readthedocs.io/en/stable/protocols.html).

**Comment by JelleZijlstra** [comment]
Sequence is not in fact a Protocol; I think that choice was made because it has too many methods.

Haven't looked too deeply into this but I suspect we can fix this in typeshed by making `reversed()` accept a custom protocol with `__getitem__` and `__len__`. I'll transfer this issue to typeshed.
